monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/MergePrioritizedListObservable.scala

```diff
@@ -0,0 +1,194 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package monix.reactive.internal.builders
+
+import monix.execution.Ack.{Continue, Stop}
+import monix.execution.cancelables.CompositeCancelable
+import monix.execution.{Ack, Cancelable, Scheduler}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.collection.mutable
+import scala.concurrent.{Future, Promise}
+import scala.util.Success
+import scala.util.control.NonFatal
+
+/** Given a sequence of priority/observable pairs, combines them into a new
+  * observable that eagerly emits source items downstream as soon as demand is
+  * signaled, choosing the item from the highest priority (greater numbers
+  * mean higher priority) source when items from multiple sources are
+  * available. If items are available from multiple sources with the same
+  * highest priority, one of them is chosen arbitrarily.
+  *
+  * Source items are buffered only to the extent necessary to accommodate
+  * backpressure from downstream, and thus if only a single item is available
+  * when demand is signaled, it will be emitted regardless of priority.
+  *
+  * Backpressure is propagated from downstream to the source observables, so
+  * that items from a given source will always be emitted downstream in the
+  * same order as received from the source, and at most a single item from a
+  * given source will be in flight at a time.
+  */
+private[reactive] final class MergePrioritizedListObservable[A](
+    sources: Seq[(Int, Observable[A])])
+    extends Observable[A] {
+
+  override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    import out.scheduler
+
+    val numberOfObservables = sources.size
+
+    val lock = new AnyRef
+    var isDone = false
+
+    // MUST BE synchronized by `lock`
+    var lastAck = Continue: Future[Ack]
+
+    case class PQElem(data: A, promise: Promise[Ack], priority: Int)
+
+    object PQElemOrdering extends Ordering[PQElem] {
+      override def compare(x: PQElem, y: PQElem): Int =
+        x.priority.compareTo(y.priority)
+    }
+
+    val pq = new mutable.PriorityQueue[PQElem]()(PQElemOrdering)
+
+    // MUST BE synchronized by `lock`
+    var completedCount = 0
+
+    // MUST BE synchronized by `lock`
+    def rawOnNext(a: A): Future[Ack] = {
+      if (isDone) Stop
+      else {
+        out
+          .onNext(a)
+          .map {
+            case Stop =>
+              lock.synchronized(completePromises())
+              Stop
+            case Continue => Continue
+          }
+          .recover {
+            case NonFatal(e) =>
+              lock.synchronized(completePromises())
+              throw e
+          }
+      }
+    }
```

Comment:
Added the above logic to ensure that we complete the upstream onNext promises when the downstream stops or errors early while we still have pending upstream items. Added tests to verify as well.

Reply:
We could use our `syncOnStopOrFailure` syntax here, should be more efficient

Reply:
Much nicer, thanks.